core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java

```diff
@@ -254,11 +254,65 @@ void validateAndApplyFeeAndQuote(
                 "Quote asset is [%s] different from the transaction sending asset [%s]",
                 quote.getSellAmount(), txn.getAmountIn()));
       }
+    }
+  }
+
+  void queryFee() throws AnchorException {
+    Sep31PostTransactionRequest request = Context.get().getRequest();
+    JwtToken token = Context.get().getJwtToken();
+    org.stellar.platform.apis.shared.Amount fee =
+        feeIntegration
+            .getFee(
+                GetFeeRequest.builder()
+                    .sendAmount(request.getAmount())
+                    .sendAsset(request.getAssetCode())
+                    .receiveAsset(request.getAssetCode())
+                    .receiveAmount(null)
+                    .senderId(request.getSenderId())
+                    .receiverId(request.getReceiverId())
+                    .clientDomain(token.getClientDomain())
+                    .build())
+            .getFee();
+
+    Context.get().setFee(fee);
+  }
+
+  void applyFeeAndQuote() throws AnchorException {
+    Sep31Transaction txn = Context.get().getTransaction();
+    Sep31PostTransactionRequest request = Context.get().getRequest();
+    Amount fee = Context.get().getFee();
+
+    AssetInfo asset = assetService.getAsset(request.getAssetCode(), request.getAssetIssuer());
+    BigDecimal amountIn = decimal(txn.getAmountIn(), asset);
+    BigDecimal amountFee = decimal(fee.getAmount(), asset);
+    if (request.getQuoteId() == null) {
+      // No quote. Apply fee.
+      // Set amount_fee
+      txn.setAmountFee(String.valueOf(amountFee));
+      txn.setAmountFeeAsset(txn.getAmountInAsset());
+      // amount_out = amount_in - amount_fee
+      BigDecimal amountOut = amountIn.subtract(amountFee);
+      txn.setAmountOut(String.valueOf(amountOut));
+      txn.setAmountOutAsset(txn.getAmountInAsset());
+    } else {
+      // Quote exists
       txn.setQuoteId(request.getQuoteId());
+      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
+      BigDecimal buyAmount = decimal(quote.getBuyAmount(), asset);
+      if (fee.getAsset().equals(txn.getAmountInAsset())) {
+        // Fee is applied to the amount_in
+        // amount_out = quote.buy_amount
+        txn.setAmountOut(String.valueOf(buyAmount));
```

Comment:
If fee is applied to `amount_in`, `amount_out` will still be less than `buy_amount`, unless `additive` is specified.